src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAdminController.java

```diff
@@ -143,17 +168,99 @@ public void fullDeleteStudy(Long[] ids) {
+            return result;
+        }
+
+        final Set<Tenant> tenants = new HashSet<>(tenantService.findByIdsIn(bulkUsersDto.getTenantIds()));
+        final boolean emailConfirmationRequired = bulkUsersDto.getEmailConfirmationRequired();
+        List<AdminUserDTO> createdUsers = bulkUsersDto.getUsers().stream()
+                .map(userDto -> {
+                    U user = convert(userDto);
+                    updateFields(user, tenants, emailConfirmationRequired, bulkUsersDto.getPassword());
+                    U createdUser = createUserAndSetValidationError(user, result, errorMessages);
+                    sendRegistrationEmail(emailConfirmationRequired, userDto, createdUser);
+                    return createdUser;
+                })
+                .map(createdUser -> conversionService.convert(createdUser, AdminUserDTO.class))
+                .collect(Collectors.toList());
+
+        result.setResult(createdUsers);
+        if (result.getErrorCode().equals(VALIDATION_ERROR.getCode())) {
+            result.setValidatorErrors(errorMessages);
+        }
+
+        return result;
+    }
+
+    private U createUserAndSetValidationError(U user, JsonResult<List<AdminUserDTO>> result, Map<String, Object> errorMessages) {
+
+        U createdUser = null;
+        try {
+            createdUser = userService.create(user);
+        } catch (Exception ex) {
+            LOGGER.error("Failed to save user with " + user.getEmail(), ex);
+            result.setErrorCode(VALIDATION_ERROR.getCode());
+            errorMessages.put(String.valueOf(user.getId()), ex.getMessage());
+        }
+
+        return createdUser;
+    }
+
+    protected abstract U convert(CommonUserRegistrationDTO dto);
+
+    private void sendRegistrationEmail(
+            final boolean emailConfirmationRequired,
+            final CommonUserRegistrationDTO userDto,
+            final U createdUser) {
+        if (emailConfirmationRequired) {
+            userService.sendRegistrationEmail(createdUser, userDto.getRegistrantToken(), userDto.getCallbackUrl(), true);
+        }
+    }
+
+    private void updateFields(U user, Set<Tenant> tenants, boolean emailConfirmationRequired, String password) {
+        user.setPassword(password);
+        user.setTenants(tenants);
+        user.setOrigin(UserOrigin.NATIVE);
+        if (!emailConfirmationRequired) {
+            user.setEmailConfirmed(true);
+        } else {
+            user.setEmailConfirmed(false);
+            user.setRegistrationCode(UUID.randomUUID().toString());
+        }
+    }
+
     @ApiOperation(value = "Get user ids.", hidden = true)
     @RequestMapping(value = "/api/v1/admin/users/ids", method = RequestMethod.GET)
     public List<String> getListOfUserIdsByFilter(final UserSearch userSearch)
             throws UserNotFoundException {
-        
-        final List<IUser> users = userService.getList(userSearch);
+  
+        final List<U> users = userService.getList(userSearch);
```

Comment:
Is getList a new method? Anyway I believe method names like getList, getPage looks uninformative.